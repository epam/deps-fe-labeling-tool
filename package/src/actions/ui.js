import { createAction } from 'redux-actions'

const FEATURE_NAME = 'UI'

const setMarkupObjectsFilter = createAction(
  `${FEATURE_NAME}/SET_MARKUP_OBJECTS_FILTER`
)

const setAssignToFieldsFilter = createAction(
  `${FEATURE_NAME}/SET_ASSIGN_TO_FIELDS_FILTER`
)

const setExpandedListKeys = createAction(`${FEATURE_NAME}/SET_KEYS`)

const toggleExpandedListKey = createAction(`${FEATURE_NAME}/TOGGLE_KEY`)
const setActiveSidebar = createAction(`${FEATURE_NAME}/SET_ACTIVE_SIDEBAR`)

const addTemporaryFieldIndex = createAction(`${FEATURE_NAME}/ADD_TEMPORARY_FIELD`)
const deleteTemporaryFieldIndex = createAction(`${FEATURE_NAME}/DELETE_TEMPORARY_FIELD_INDEX`)

const resetDefault = createAction(`${FEATURE_NAME}/RESET_DEFAULT`)

export {
  setMarkupObjectsFilter,
  setActiveSidebar,
  setExpandedListKeys,
  toggleExpandedListKey,
  addTemporaryFieldIndex,
  setAssignToFieldsFilter,
  deleteTemporaryFieldIndex,
  resetDefault
}
