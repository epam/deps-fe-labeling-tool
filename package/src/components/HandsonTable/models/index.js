
export * from './HTTableData'
export * from './HTMerge'
export * from './HTColumn'
