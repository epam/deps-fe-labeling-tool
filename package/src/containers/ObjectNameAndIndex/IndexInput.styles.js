import styled from 'styled-components'
import { Input } from '@/components/Input'
import { COLORS } from '@/theme/theme.default'

const StyledNumericInput = styled(Input.Numeric)`
  flex: 0 1 6.4rem;
  background-color: ${COLORS.PRIMARY_3};

  & input {
    padding: 0 0.8rem;
  }

  & input:disabled{
    background-color: ${COLORS.PRIMARY_3};
  }
`

export {
  StyledNumericInput as NumericInput
}
