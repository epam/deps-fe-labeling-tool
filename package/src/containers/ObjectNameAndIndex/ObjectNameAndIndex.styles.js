import styled from 'styled-components'
import { Typography } from '@/components/Typography'
import { COLORS } from '@/theme/theme.default'

const { Paragraph } = Typography

const ObjectName = styled(Paragraph)`
  max-width: 190px;
  height: 2.4rem;
  font-size: 1.2rem;
  line-height: 1.6rem;
  color: ${COLORS.PRIMARY_4};
  padding: 0.4rem 0 0 0.6rem;
  margin-right: 1rem;
  border-left: 2px solid ${(props) => props.color};
`

const ObjectType = styled.span`
  font-size: 1rem;
  line-height: 1.2rem;
  color: ${COLORS.GRAYSCALE_5};
`

const Wrapper = styled.div`
  display: flex;
  justify-content: space-between;
  height: 2.4rem;
`

export {
  ObjectName,
  ObjectType,
  Wrapper
}
