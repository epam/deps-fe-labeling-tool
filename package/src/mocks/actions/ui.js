const setMarkupObjectsFilter = jest.fn((filter) => ({
  type: 'MOCK_MARKUP_OBJECTS_FILTER_ACTION',
  payload: filter
}))

const setAssignToFieldsFilter = jest.fn((filter) => ({
  type: 'MOCK_ASSIGN_TO_FIELDS_FILTER_ACTION',
  payload: filter
}))

const setExpandedListKeys = jest.fn((key) => ({
  type: 'MOCK_SET_EXPANDED_LIST_KEYS_ACTION',
  payload: key
}))

const toggleExpandedListKey = jest.fn((key) => ({
  type: 'MOCK_TOGGLE_EXPANDED_LIST_KEY_ACTION',
  payload: key
}))

const addTemporaryFieldIndex = jest.fn((key) => ({
  type: 'MOCK_ADD_TEMPORARY_FIELD_INDEX',
  payload: key
}))

const deleteTemporaryFieldIndex = jest.fn((key) => ({
  type: 'MOCK_DELETE_TEMPORARY_FIELD_INDEX',
  payload: key
}))

const resetDefault = jest.fn(() => ({
  type: 'MOCK_RESET_DEFAULT'
}))

const setActiveSidebar = jest.fn(() => ({
  type: 'MOCK_SET_ACTIVE_SIDEBAR'
}))

const mockUiActions = {
  setExpandedListKeys,
  toggleExpandedListKey,
  setMarkupObjectsFilter,
  setAssignToFieldsFilter,
  addTemporaryFieldIndex,
  deleteTemporaryFieldIndex,
  resetDefault,
  setActiveSidebar
}

export {
  mockUiActions
}
