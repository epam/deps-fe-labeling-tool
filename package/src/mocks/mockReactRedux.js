const mockDispatch = jest.fn((action) => action)

const createMockMapStateToProps = (mapStateToProps) => {
  if (!mapStateToProps) {
    return mapStateToProps
  }

  return (state, ownProps) => ({
    props: mapStateToProps(state, ownProps)
  })
}

const getDispatchFunctionFormShorthandObject = (actionsObj) => (dispatch) => (
  Object.keys(actionsObj).reduce(
    (props, actionKey) => ({
      ...props,
      [actionKey]: jest.fn(
        (...args) => dispatch(actionsObj[actionKey].apply(undefined, args))
      )
    }),
    {}
  )
)

const createMockMapDispatchToProps = (mapDispatchToProps) => {
  if (!mapDispatchToProps) {
    return mapDispatchToProps
  }

  const mappingFunction = typeof mapDispatchToProps === 'function'
    ? mapDispatchToProps
    : getDispatchFunctionFormShorthandObject(mapDispatchToProps)

  return (dispatch, ownProps) => {
    dispatch = dispatch || mockDispatch

    return {
      props: mappingFunction(dispatch, ownProps),
      dispatch
    }
  }
}

const createMockMergeProps = (mergeProps) => {
  if (!mergeProps) {
    return mergeProps
  }

  return (stateProps, dispatchProps, ownProps) => {
    dispatchProps = dispatchProps || { dispatch: mockDispatch }

    return {
      props: mergeProps(stateProps, dispatchProps, ownProps),
      dispatch: dispatchProps.dispatch
    }
  }
}

const mockReactRedux = {
  connect: (mapStateToProps, mapDispatchToProps, mergeProps) => (WrappedComponent) => ({
    mapStateToProps: createMockMapStateToProps(mapStateToProps),
    mapDispatchToProps: createMockMapDispatchToProps(mapDispatchToProps),
    mergeProps: createMockMergeProps(mergeProps),
    WrappedComponent
  }),
  batch: (fn) => Promise.resolve(fn),
  useDispatch: jest.fn(() => mockDispatch),
  useSelector: jest.fn((selector) => selector())
}

export {
  mockReactRedux
}
