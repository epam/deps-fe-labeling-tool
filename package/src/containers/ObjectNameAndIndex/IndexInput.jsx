import React, { useCallback } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { deleteTemporaryFieldIndex } from '@/actions/ui'
import { labelShape } from '@/models/Label'
import { tableShape } from '@/models/Table'
import { NumericInput } from './IndexInput.styles'

const INDEX_MIN_VALUE = 1

const IndexInput = ({
  onChange,
  object,
  className,
  deleteTemporaryFieldIndex
}) => {
  const onIndexChange = useCallback(
    (value) => {
      if (object.index === value) {
        return
      }

      deleteTemporaryFieldIndex({
        code: object.fieldCode,
        index: value
      })

      onChange(
        {
          ...object,
          index: value
        }
      )
    },
    [
      object,
      onChange,
      deleteTemporaryFieldIndex
    ]
  )

  return (
    <NumericInput
      className={className}
      min={INDEX_MIN_VALUE}
      value={object.index}
      onChange={onIndexChange}
    />
  )
}

IndexInput.propTypes = {
  className: PropTypes.string,
  deleteTemporaryFieldIndex: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  object: PropTypes.oneOfType([
    labelShape,
    tableShape
  ]).isRequired
}

const mapDispatchToProps = {
  deleteTemporaryFieldIndex
}

const ConnectedComponent = connect(null, mapDispatchToProps)(IndexInput)

export {
  ConnectedComponent as IndexInput
}
