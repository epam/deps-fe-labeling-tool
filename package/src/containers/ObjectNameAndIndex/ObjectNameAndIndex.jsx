import React, { useMemo } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Tooltip } from '@/components/Tooltip'
import { FieldType } from '@/enums/FieldType'
import { fieldShape } from '@/models/Field'
import { labelShape, LabelType } from '@/models/Label'
import { Markup, markupShape } from '@/models/Markup'
import { tableShape, TABLE_TYPE_NAME } from '@/models/Table'
import { markupSelector } from '@/selectors/markup'
import { fieldsSelector } from '@/selectors/model'
import { COLORS } from '@/theme/theme.default'
import { IndexInput } from './IndexInput'
import {
  ObjectName,
  Wrapper,
  ObjectType
} from './ObjectNameAndIndex.styles'

const PAIR_NAME_SEPARATOR = '.'
const RESOURCE_UNASSIGNED = 'Unassigned'

const OBJECT_TYPE_TO_COLOR = {
  [LabelType.CHECKMARK]: COLORS.LABEL_CHECKMARK_BORDER,
  [LabelType.KEY]: COLORS.LABEL_KEY_BORDER,
  [LabelType.VALUE]: COLORS.LABEL_VALUE_BORDER,
  [LabelType.STRING]: COLORS.LABEL_STRING_BORDER,
  [LabelType.ENUM]: COLORS.LABEL_ENUM_BORDER,
  [LabelType.DATE]: COLORS.LABEL_DATE_BORDER,
  [TABLE_TYPE_NAME]: COLORS.TABLE_BORDER
}

const ObjectNameAndIndex = ({ object, onChange, fields, markup }) => {
  const indexAvailable = useMemo(
    () => {
      const { fieldType } = fields.find((f) => f.code === object.fieldCode) ?? {}
      return fieldType === FieldType.LIST
    },
    [object, fields]
  )

  const getKeyValuePairName = (name, type) => (
    <span>
      {`${name}${PAIR_NAME_SEPARATOR}`}
      <ObjectType>
        {type}
      </ObjectType>
    </span>
  )

  const getObjectColor = () => {
    if (!object.fieldCode) {
      return COLORS.MARKUP_OBJECT_UNASSIGNED_BORDER
    }

    return OBJECT_TYPE_TO_COLOR[object.type ?? object.typeName]
  }

  const getUnassignedObjectName = () => {
    const allMarkupObjects = Markup.getAllObjects(markup)

    const filteredMarkupObjects = allMarkupObjects.filter((markupObject) => (
      !markupObject.fieldCode &&
      markupObject.typeName === object.typeName
    ))

    const objectIndex = filteredMarkupObjects.findIndex((markupObject) => (
      markupObject.uid === object.uid
    ))

    return `${RESOURCE_UNASSIGNED} ${object.typeName} ${objectIndex + 1}`
  }

  const getObjectName = () => {
    const objectName = fields.find((field) => field.code === object.fieldCode)?.name

    if (object.fieldCode === '') {
      return getUnassignedObjectName()
    }

    if (object.type === LabelType.KEY || object.type === LabelType.VALUE) {
      return getKeyValuePairName(objectName, object.type)
    }

    return objectName
  }

  return (
    <Wrapper>
      <ObjectName
        ellipsis={
          {
            tooltip: (
              <Tooltip
                title={object.typeName}
              >
                {getObjectName()}
              </Tooltip>
            )
          }
        }
        color={getObjectColor()}
      >
        {getObjectName()}
      </ObjectName>
      {
        indexAvailable &&
        (
          <IndexInput
            onChange={onChange}
            object={object}
            fields={fields}
          />
        )
      }
    </Wrapper>
  )
}

ObjectNameAndIndex.propTypes = {
  object: PropTypes.oneOfType([
    labelShape, tableShape
  ]).isRequired,
  onChange: PropTypes.func.isRequired,
  fields: PropTypes.arrayOf(fieldShape).isRequired,
  markup: markupShape.isRequired
}

const mapStateToProps = (state) => ({
  fields: fieldsSelector(state),
  markup: markupSelector(state)
})

const ConnectedComponent = connect(mapStateToProps)(ObjectNameAndIndex)

export {
  ConnectedComponent as ObjectNameAndIndex
}
