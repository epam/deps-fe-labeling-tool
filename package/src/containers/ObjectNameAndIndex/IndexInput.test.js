import { mockUiActions } from '@/mocks/actions/ui'
import { mockReactRedux } from '@/mocks/mockReactRedux'
import React from 'react'
import { shallow } from 'enzyme'
import { deleteTemporaryFieldIndex } from '@/actions/ui'
import { IndexInput } from '@/containers/ObjectNameAndIndex/IndexInput'
import { Field } from '@/models/Field'
import { Table } from '@/models/Table'
import { NumericInput } from './IndexInput.styles'

jest.mock('react-redux', () => mockReactRedux)
jest.mock('@/actions/ui', () => mockUiActions)

const { WrappedComponent, mapDispatchToProps } = IndexInput

const fields = [
  new Field('1', 'FirstField', 'string'),
  new Field('2', 'SecondField', 'pair'),
  new Field('3', 'ThirdField', 'table')
]

const testTable = new Table(
  [790, 1190],
  [550, 580],
  [],
  [],
  { meta: 'some meta' },
  'countryState',
  1
)
const mockOnChange = jest.fn()

describe('Container: IndexInput', () => {
  describe('mapDispatchToProps', () => {
    const { props } = mapDispatchToProps()

    it('should pass deleteTemporaryFieldIndex action as prop to WrappedComponent', () => {
      props.deleteTemporaryFieldIndex({
        code: 'test',
        index: 1
      })
      expect(deleteTemporaryFieldIndex).nthCalledWith(1, {
        code: 'test',
        index: 1
      })
    })
  })

  describe('WrappedComponent', () => {
    let wrapper, defaultProps

    beforeEach(() => {
      defaultProps = {
        ...mapDispatchToProps().props,
        fields,
        onChange: mockOnChange,
        object: testTable
      }
      wrapper = shallow(<WrappedComponent {...defaultProps} />)
      jest.clearAllMocks()
    })

    it('should render layout correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should call onChange prop when index from table and value from arg do not match', () => {
      const input = wrapper.find(NumericInput)
      input.props().onChange('nomatch')
      expect(defaultProps.onChange).toHaveBeenCalledTimes(1)
    })

    it('should not call onChange prop when index from table and value arg match', () => {
      const input = wrapper.find(NumericInput)
      input.props().onChange(1)
      expect(defaultProps.onChange).toHaveBeenCalledTimes(0)
    })

    it('should call deleteTemporaryIndex when change index', () => {
      const mockNewIndex = 3
      const input = wrapper.find(NumericInput)
      input.props().onChange(mockNewIndex)
      expect(deleteTemporaryFieldIndex).nthCalledWith(1, {
        code: defaultProps.object.fieldCode,
        index: mockNewIndex
      })
    })
  })
})
