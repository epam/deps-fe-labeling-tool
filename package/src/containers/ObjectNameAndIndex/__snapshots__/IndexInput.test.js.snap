// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container: IndexInput WrappedComponent should render layout correctly 1`] = `
<IndexInputstyles__StyledNumericInput
  min={1}
  onChange={[Function]}
  value={1}
/>
`;
